apps/src/netsim/NetSimLobby.js

```diff
@@ -236,31 +236,45 @@ NetSimLobby.prototype.setShardID = function (shardID) {
  */
 NetSimLobby.prototype.onShardChange_ = function (shard, myNode) {
   // Unregister old handlers
-  if (this.eventKeys.nodeTable) {
-    this.shard_.nodeTable.tableChange.unregister(this.eventKeys.nodeTable);
-  }
-  if (this.eventKeys.wireTable) {
-    this.shard_.wireTable.tableChange.unregister(this.eventKeys.wireTable);
+  if (this.eventKeys.registeredShard) {
+    this.eventKeys.registeredShard.nodeTable.tableChange.unregister(
+        this.eventKeys.nodeTable);
+    this.eventKeys.registeredShard.wireTable.tableChange.unregister(
+        this.eventKeys.wireTable);
+    this.registeredShard = undefined;
   }
 
   this.shard_ = shard;
   this.myNode_ = myNode;
 
-  if (!this.shard_) {
-    // If we disconnected, just clear our lobby data
+  if (this.shard_) {
+    // We got connected to a shard!
+    // Register for events
+    this.eventKeys.nodeTable = this.shard_.nodeTable.tableChange.register(
+        this.onNodeTableChange_.bind(this));
+    this.eventKeys.wireTable = this.shard_.wireTable.tableChange.register(
+        this.onWireTableChange_.bind(this));
+    this.eventKeys.registeredShard = this.shard_;
+
+    // Trigger a forced read of the node table
+    this.fetchInitialLobbyData_();
+  } else {
+    // We've been disconnected from a shard
+    // Clear our selected shard ID
+    this.selectedShardID_ = undefined;
+
+    // Clear cached lobby data
     this.nodesOnShard_.length = 0;
     this.incomingConnectionNodes_.length = 0;
-    return;
-  }
 
-  // Register for events
-  this.eventKeys.nodeTable = this.shard_.nodeTable.tableChange.register(
-      this.onNodeTableChange_.bind(this));
-  this.eventKeys.wireTable = this.shard_.wireTable.tableChange.register(
-      this.onWireTableChange_.bind(this));
+    // Redraw the lobby
+    this.render();
 
-  // Trigger a forced read of the node table
-  this.fetchInitialLobbyData_();
+    // If there's only one option, try to auto-reconnect
+    if (this.shardChoices_.length === 1) {
+      this.setShardID(this.shardChoices_[0].shardID);
+    }
```

Comment:
Three new things the lobby does when it detects a disconnect:
1. It clears the selectedShardID_ (we've been kicked, we don't have one now)
2. It triggers a `render()`; in the connect path this happens as a result of `fetchInitialLobbyData_()` but it was getting missed on disconnect
3. If there's only one possible shard anyway, trigger an auto-reconnect to that shard (which puts the student back in the "Pick a Peer/Router" screen)